WordPress/Classes/Services/PostService.m

```diff
@@ -333,70 +337,137 @@ - (void)autoSave:(AbstractPost *)post
+}
+
+- (void)autoSave:(AbstractPost *)post
+         success:(void (^)(AbstractPost *post, NSString *previewURL))success
+         failure:(void (^)(NSError * _Nullable error))failure
+{
+    id<PostServiceRemote> remote = [self.postServiceRemoteFactory forBlog:post.blog];
+    if ([remote isKindOfClass:[PostServiceRemoteREST class]]) {
+        [self handleAutoSaveWithRestRemote:(PostServiceRemoteREST *)remote forPost:post success:success failure:failure];
 
-            [self uploadPost:post
-        forceDraftIfCreating:YES
-                     success:^(AbstractPost * _Nonnull post) {
-                         success(post, nil);
-                     }
-                     failure:failure];
-        } else {
-            [restRemote autoSave:remotePost
-                         success:successBlock
-                         failure:setPostAsFailedAndCallFailureBlock];
-        }
     } else if ([post originalIsDraft] && [post isDraft]) {
         [self uploadPost:post
                  success:^(AbstractPost * _Nonnull post) {
                      success(post, nil);
                  } failure:failure];
+
     } else {
-        setPostAsFailedAndCallFailureBlock(defaultError);
+        NSError *defaultError = [self defaultAutosaveError];
+        AutosaveFailureBlock failureBlock = [self wrappedAutosaveFailureBlock:post failure:failure];
+        failureBlock(defaultError);
     }
 }
 
+- (void)handleAutoSaveWithRestRemote:(PostServiceRemoteREST *)restRemote
+                             forPost:(AbstractPost *)post
+                             success:(void (^)(AbstractPost *post, NSString *previewURL))success
+                             failure:(void (^)(NSError * _Nullable error))failure
+{
+    NSManagedObjectID *postObjectID = post.objectID;
+    RemotePost *remotePost = [self remotePostWithPost:post];
+
+    AutosaveSuccessBlock autosaveSuccessBlock = [self wrappedAutosaveSuccessBlock:postObjectID success:success];
+    AutosaveFailureBlock setPostAsFailedAndCallFailureBlock = [self wrappedAutosaveFailureBlock:post failure:failure];
+
+    // The autoSave endpoint returns an exception on posts that do not exist on the server
+    // so we'll create the post instead if necessary.
+    BOOL mustBeCreated = ![post hasRemote];
+
+    if (mustBeCreated) {
+        // Abort if the status is trashed/deleted. We'd rather not automatically create a
+        // locally trashed post as drafts in the server.
+        if ([post.status isEqualToString:PostStatusTrash] || [post.status isEqualToString:PostStatusDeleted]) {
+            NSError *error = [self defaultAutosaveError];
+            setPostAsFailedAndCallFailureBlock(error);
+            return;
+        }
+
+        [self uploadPost:post
+    forceDraftIfCreating:YES
+                 success:^(AbstractPost * _Nonnull post) {
+                     success(post, nil);
+                 }
+                 failure:failure];
+        return;
+    }
+
+    if ([post originalIsDraft]) {
```

Comment:
⚠️ I'm sorry I didn't see this before. I found another scenario where the comments can still be disabled. It happens in the offline posting cancellation flow:

1. Create a draft and upload it. 
2. Go offline. 
3. Still in the Post List, tap More → Publish Now. The upload should fail and there should be a Cancel button visible. 
4. Tap on Cancel button. 
5. Go online.

Going online will trigger auto-uploading. However, the **explicit cancelation** of the user will disallow "uploads". In cases like this, we still try to save the user's changes by autosaving instead. This is where that decision happens:

https://github.com/wordpress-mobile/WordPress-iOS/blob/c957f3d86b4a4481ca8e5cc3b0c3e20473a4f810/WordPress/Classes/Services/PostAutoUploadInteractor.swift#L37-L63

So, eventually, `PostService.handleAutoSaveWithRestRemote` will still be called. The problem is the **`AbstractPost.status` will still be `.publish`** at this point. This `[post originalIsDraft]` condition will **fail** and the routine will move to line 463 which causes the comment-disabling bug again:

https://github.com/wordpress-mobile/WordPress-iOS/blob/894bd2cf8c8477e1c963a0dfaeb94152a189d248/WordPress/Classes/Services/PostService.m#L462-L465

I think we can get rid of this `[post originalIsDraft]` and condition and always attempt to do a `getPostWithID` before calling `autosave`. And also get rid of the `autosave` call in line 463.

Reply:
Good catch! The tweak makes sense and it seems the safest thing to always check the current remote value of status.